src/util/test_util/mock_vm.rs

```diff
@@ -0,0 +1,597 @@
+            })),
+
+            stop_all_mutators: MockMethod::new_unimplemented(),
+            resume_mutators: MockMethod::new_unimplemented(),
+            block_for_gc: MockMethod::new_unimplemented(),
+            spawn_gc_thread: MockMethod::new_default(),
+            out_of_memory: MockMethod::new_fixed(Box::new(|(_, err)| {
+                panic!("Out of memory with {:?}!", err)
+            })),
+            schedule_finalization: MockMethod::new_default(),
+            post_forwarding: MockMethod::new_default(),
+            vm_live_bytes: MockMethod::new_default(),
+
+            copy_object: MockMethod::new_unimplemented(),
+            copy_object_to: MockMethod::new_unimplemented(),
+            get_object_size: MockMethod::new_unimplemented(),
+            get_object_size_when_copied: MockMethod::new_unimplemented(),
+            get_object_align_when_copied: MockMethod::new_fixed(Box::new(|_| {
+                std::mem::size_of::<usize>()
+            })),
+            get_object_align_offset_when_copied: MockMethod::new_fixed(Box::new(|_| 0)),
+            get_type_descriptor: MockMethod::new_unimplemented(),
+            get_object_reference_when_copied_to: MockMethod::new_unimplemented(),
+            ref_to_object_start: MockMethod::new_fixed(Box::new(|object| {
+                object.to_raw_address().sub(DEFAULT_OBJECT_REF_OFFSET)
+            })),
+            ref_to_header: MockMethod::new_fixed(Box::new(|object| object.to_raw_address())),
+            ref_to_address: MockMethod::new_fixed(Box::new(|object| {
+                object.to_raw_address().sub(DEFAULT_OBJECT_REF_OFFSET)
+            })),
+            address_to_ref: MockMethod::new_fixed(Box::new(|addr| {
+                ObjectReference::from_raw_address(addr.add(DEFAULT_OBJECT_REF_OFFSET))
+            })),
+            dump_object: MockMethod::new_unimplemented(),
+
+            weakref_clear_referent: MockMethod::new_unimplemented(),
+            weakref_get_referent: MockMethod::new_unimplemented(),
+            weakref_set_referent: MockMethod::new_unimplemented(),
+            weakref_is_referent_cleared: MockMethod::new_fixed(Box::new(|r| r.is_null())),
+            weakref_enqueue_references: MockMethod::new_unimplemented(),
+
+            support_edge_enqueuing: MockMethod::new_fixed(Box::new(|_| true)),
+            scan_object: MockMethod::new_unimplemented(),
+            scan_object_and_trace_edges: MockMethod::new_unimplemented(),
+            scan_roots_in_mutator_thread: Box::new(MockMethod::<
+                (
+                    VMWorkerThread,
+                    &'static mut Mutator<MockVM>,
+                    ProcessEdgesWorkRootsWorkFactory<
+                        MockVM,
+                        SFTProcessEdges<MockVM>,
+                        SFTProcessEdges<MockVM>,
+                    >,
+                ),
+                (),
+            >::new_unimplemented()),
```

Comment:
This is one example of API methods that have deep call-back structures.  The call sequence of a normal VM is:

1.  MMTk core wants to scan the roots of a Mutator.  It calls `scan_roots_in_mutator_thread`, passing an `impl RootsWorkFactory` instance.
2.  The VM binding implements this method.  It calls methods of the `RootsWorkFactory` to deliver roots.  For example, it calls `create_process_edge_roots_work` to deliver a list of slots (`Vec<impl Edge>`)

The testing script needs to attach a closure to `scan_roots_in_mutator_thread` when it is testing, for example, the `ScanMutatorRoots<C>` work packet.  However, the `ScanMutatorRoots<C>` work packet has a type parameter `<C>` and will determine the concrete `RootsWorkFactory` implementation.  Therefore, a hard-coded type

```
ProcessEdgesWorkRootsWorkFactory<
                        MockVM,
                        SFTProcessEdges<MockVM>,
                        SFTProcessEdges<MockVM>,
                    >
```

will simply not work.  The `MockAny` trait hides the concrete type, but the downcast will be guaranteed to fail (unless the chosen `Plan` actually uses `SFTProcessEdges` for both the `E` and the `I` type arguments of `RootsWorkFactory`, but we use `PlanProcessEdges` most of the time).

To solve the problem, we should look at what the `RootsWorkFactory` really is. It is really just a fancy way for the VM binding to deliver lists of roots to MMTk core.  So when we design the mocking facility, we can be more straightforward about this -- just let the closure deliver the roots.

Add a new enum type:

```rs
pub enum RootsWorkFactoryCall {
    Slots(Vec<MockVMEdge>),
    PinNodes(Vec<ObjectReference>),
    TPinNodes(Vec<ObjectReference>),
}
```

Change the `scan_roots_in_mutator_thread` and `scan_vm_specific_roots` fields to a `MockMethod`:

```rs
    pub scan_roots_in_mutator_thread: MockMethod<(VMWorkerThread, &'static mut Mutator<MockVM>), Vec<RootsWorkFactoryCall>>,
    pub scan_vm_specific_roots: MockMethod<VMWorkerThread, Vec<RootsWorkFactoryCall>>,
```

Note that we remove the `RootsWorkFactory` from the arguments, but add a `Vec<RootsWorkFactoryCall>` as the return value.

Then when implementing those methods for `Scanning`, we call the closure, and then call the actual factory according to the return values:

```rs
    fn scan_roots_in_mutator_thread(
        tls: VMWorkerThread,
        mutator: &'static mut Mutator<Self>,
        mut factory: impl RootsWorkFactory<<MockVM as VMBinding>::VMEdge>,
    ) {
        let results = mock!(scan_roots_in_mutator_thread(tls, mutator));
        for result in results {
            match result {
                RootsWorkFactoryCall::Slots(edges) => factory.create_process_edge_roots_work(edges),
                RootsWorkFactoryCall::PinNodes(nodes) => factory.create_process_pinning_roots_work(nodes),
                RootsWorkFactoryCall::TPinNodes(nodes) => factory.create_process_tpinning_roots_work(nodes),
            }
        }
    }
    fn scan_vm_specific_roots(
        tls: VMWorkerThread,
        mut factory: impl RootsWorkFactory<<MockVM as VMBinding>::VMEdge>,
    ) {
        let results = mock!(scan_vm_specific_roots(tls));
        for result in results {
            match result {
                RootsWorkFactoryCall::Slots(edges) => factory.create_process_edge_roots_work(edges),
                RootsWorkFactoryCall::PinNodes(nodes) => factory.create_process_pinning_roots_work(nodes),
                RootsWorkFactoryCall::TPinNodes(nodes) => factory.create_process_tpinning_roots_work(nodes),
            }
        }
    }
```

Reply:
> Therefore, a hard-coded type will simply not work. The `MockAny` trait hides the concrete type, but the downcast will be guaranteed to fail (unless the chosen `Plan` actually uses `SFTProcessEdges` for both the `E` and the `I` type arguments of `RootsWorkFactory`, but we use `PlanProcessEdges` most of the time).

Right. The default `MockMethod` is just a place holder and it won't work, as it carries some type information that won't fit in some cases. I used `SFTProcessEdges` in the place holder as it is simple to write. I will probably make this clearer. If someone wants to mock this method, they should create a `MockMethod` for their tests. For example,
```rust
MockVM {
  scan_roots_in_mutator_thread: Box::new(MockMethod::<..., ProcessEdgesWorkRootsWorkFactory<MockVM, PlanProcessEdges<MockVM, Immix, DEFAULT_TRACE>, PlanProcessEdges<MockVM, Immix, DEFAULT_TRACE>>, ()>,
  ..MockVM::default()
}
```
Do you think this would work?

But as you said, it only works for one plan, which means the user needs to make sure the tests only run for the plan. However, that is how our code works right now, rather than an issue of the mocking.

> Change the `scan_roots_in_mutator_thread` and `scan_vm_specific_roots` fields to a `MockMethod`:
> 
> ```rust
>     pub scan_roots_in_mutator_thread: MockMethod<(VMWorkerThread, &'static mut Mutator<MockVM>), Vec<RootsWorkFactoryCall>>,
>     pub scan_vm_specific_roots: MockMethod<VMWorkerThread, Vec<RootsWorkFactoryCall>>,
> ```

I don't like to transform some methods for mocking in an ad-hoc way. I hope we can mock any method (including new methods) by following some simple rules (like what I wrote down in the comments -- we just need to get rid of generic lifetime and type parameters). If we have to deal with some methods in an ad-hoc way, it mostly means the mocking cannot handle some cases, and there is a chance that we cannot mock some methods at all.

Reply:
I have fixed other issues in the comments. Can you take a look at this? @wks I will try get it resolved.

Reply:
> Do you think this would work?

Probably.  I understand your point that the user of the MockVM can set the MockMethod on a per-instance basis, and they can always set it to the proper MockMethod that works for their particular test.

Reply:
I suggest adding a comment here, letting the reader know your point in the following paragraph

> Right. The default MockMethod is just a place holder and it won't work, as it carries some type information that won't fit in some cases. I used SFTProcessEdges in the place holder as it is simple to write. I will probably make this clearer. If someone wants to mock this method, they should create a MockMethod for their tests. For example,

Reply:
Done. I added some comments for `MockAny`, and `MockVM` in https://github.com/mmtk/mmtk-core/pull/1049/commits/9fe32eee86ce9475b5adf0302b52d3e4f658d6be. I further added some comments when we create the `MockMethod` for `MockAny` in https://github.com/mmtk/mmtk-core/pull/1049/commits/85227dfd02188f72c1d6491f52f7d015bfeaf84e.